dbms/src/Operators/AutoPassThroughHashAggHelper.cpp

```diff
@@ -0,0 +1,432 @@
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <Columns/ColumnDecimal.h>
+#include <Columns/ColumnNullable.h>
+#include <Common/Exception.h>
+#include <Common/Logger.h>
+#include <Core/Block.h>
+#include <DataTypes/DataTypeDecimal.h>
+#include <DataTypes/DataTypeNullable.h>
+#include <DataTypes/DataTypesNumber.h>
+#include <Functions/FunctionHelpers.h>
+#include <Operators/AutoPassThroughHashAggHelper.h>
+#include <common/logger_useful.h>
+
+namespace DB
+{
+// first_row and sum whose return type is same with argument type will use this generator:
+// 1. For not-null argument: init first_row result as its original value.
+// 2. For nullable argument: if original value is not null, init first_row result as its original value, Otherwise init it as Null.
+// So copy column is enough for first_row.
+ColumnPtr genPassThroughColumnByCopy(
+    const TypeIndex & required_type_id,
+    const String & src_col_name,
+    const Block & child_block)
+{
+    auto res = child_block.getByName(src_col_name);
+    RUNTIME_CHECK(required_type_id == res.type->getTypeId());
+    return res.column;
+}
+
+// For not-null argument: init count result as 1.
+// For nullable argument: init count result as 1 if row is not null, otherwise init it as zero.
+// NOTE: empty_input_as_null is not consider, because will not use AutoPassHashAgg if group by key is zero.
+template <bool nullable, bool empty_argument>
+ColumnPtr genPassThroughColumnForCount(size_t arg_index, const Block & child_block)
+{
+    auto count_agg_func_res = ColumnVector<UInt64>::create(child_block.rows(), 1);
+    if constexpr (empty_argument || !nullable)
+        return count_agg_func_res;
+
+    const ColumnWithTypeAndName & argument_column = child_block.getByPosition(arg_index);
+    const auto * col_nullable = checkAndGetColumn<ColumnNullable>(argument_column.column.get());
+    RUNTIME_CHECK(col_nullable);
+    auto & datas = count_agg_func_res->getData();
+    for (size_t i = 0; i < child_block.rows(); ++i)
+    {
+        if (col_nullable->isNullAt(i))
+            datas[i] = 0;
+    }
+    return count_agg_func_res;
+}
+
+// For not-null argument: init sum result as its original value.
+// For nullable column: if original value is not null, init sum result as its original value. Otherwise init it as Null.
+// NOTE: Also needs to handle type cast. For example arg type is decimal(15, 2), result type will be decimal(15+22, 2)
+template <typename FromDecimalType, typename ToDecimalType, bool nullable>
+ColumnPtr genPassThroughColumnForSumDecimal(size_t arg_index, const Block & child_block)
```

Comment:
todo: maybe macro for `genPassThroughColumnForSumDecimal` and `genPassThroughColumnForSumNumber`